Proofs/Experiments/IsabelleHard.lean

```diff
@@ -0,0 +1,202 @@
+/-
+Copyright (c) 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author(s): Siddharth Bhat
+
+This file contains bitvector theorems that stress Isabelle automation.
+We port the file to Lean for comparison.
+-/
+
+import Tactics.Sym
+import Init.Data.BitVec.Folds
+import Init.Data.BitVec.Bitblast
+import Init.Data.BitVec.Lemmas
+open BitVec
+
+/--
+`mask w n` is an w-bit word where the low `n`
+bits are `1`s and the upper bits are `0`s.
+ -/
+def mask (w : Nat) (i : Nat) : BitVec w :=
+  match i with
+  | 0 => 0#w
+  | i' + 1 => 1#w <<< i' ||| mask w i'
+
+@[simp]
+theorem getLsb_mask : (mask w₁ i).getLsb j =
+  (decide (j < i) && (j < w₁)) := by
+  induction i generalizing w₁ j
+  case zero => simp [mask]
+  case succ i ih =>
+    simp only [mask, getLsb_or, getLsb_shiftLeft, getLsb_one]
+    rw [ih]
+    by_cases h : j < i
+    · simp [h]; omega
+    · simp [h]
+      by_cases h' : j < w₁
+      · simp [h', show 0 < w₁ by omega]; omega
+      · simp [h']
+
+@[simp]
+def mask64 (i : Nat) : BitVec 64 := mask 64 i
+
+/--
+This is challenging: the number `n`, `m` are used both as:
+- bitvectors (for the shifts).
+- `nat`s for the mask.
+-/
+theorem mask_split (x : BitVec 64) (n m : Nat) :
+    (x &&& mask64 (n + m)) = (x &&& (mask64 m <<< n) ||| x &&& mask64 n) := by
+  simp only [mask64]
+  apply BitVec.eq_of_getLsb_eq
+  simp only [getLsb_and, getLsb_mask, Fin.is_lt, decide_True, Bool.and_true, getLsb_or,
+    getLsb_shiftLeft, Bool.true_and]
+  intros i
+  by_cases h : (i : Nat) < n
+  · simp [h]
+    omega
+  · simp [h]
+    by_cases h' : (i : Nat) < n + m
+    · simp [h']
+      omega
+    · simp [h']
+      omega
+
+/-- info: 'mask_split' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms mask_split
+
+-- ∀x, y: word64. ∀m, k: ℕ. (x & mask64 (m + k) = y) ⟹ ((x & mask64 m) = (y & mask64 m))
+theorem smaller_mask_same (x y : BitVec 64) (m k : Nat)
+    (h : x &&& mask64 (m + k) = y) :
+    x &&& mask64 m = y &&& mask64 m := by
+  apply BitVec.eq_of_getLsb_eq
+  simp
+  intros i
+  have : (x &&& mask64 (m + k)).getLsb i = (y).getLsb i := by rw [h]
+  simp only [mask64, getLsb_and, getLsb_mask, Fin.is_lt, decide_True, Bool.and_true] at this
+  by_cases h : (i : Nat) < m
+  · simpa [h, show (i : Nat) < m + k by omega] using this
+  · simp [h]
+
+/-- info: 'smaller_mask_same' depends on axioms: [propext, Classical.choice, Quot.sound] -/
+#guard_msgs in #print axioms smaller_mask_same
+
+/-
+mask_split_left
+∀x, y: word64. ∀n: ℕ.
+    ((x & (mask64 m << n) | x & mask64 n) = y)
+    ⟹
+    (x & (mask m64 << n) = (y & (mask64 m << n)))
+-/
+theorem mask_split_left (x y : BitVec 64) (n m : Nat)
+    (h : (x &&& (mask64 m <<< n) ||| x &&& mask64 n) = y) :
+    x &&& (mask64 m <<< n) = (y &&& (mask64 m <<< n)) := by
+  bv_decide  -- sus
```

Comment:
Why "sus"?

Reply:
```lean
theorem mask_split_left (x y : BitVec 64) (n m : Nat)
    (h : (x &&& (mask64 m <<< n) ||| x &&& mask64 n) = y) :
    x &&& (mask64 m <<< n) = (y &&& (mask64 m <<< n)) := by
  bv_decide  -- sus
```

I marked it as sus because I don't see how a SAT solver can prove it without unfolding what mask64 means, and IIRC LeanSAT doesn't unfold definitions. Let me understand why it succeeds.